File: android/app/src/main/kotlin/com/bockmedia/console/local/OfflineDownloadSync.kt
package com.bockmedia.console.local

import android.content.Context
import com.bockmedia.console.domain.model.PlayTarget
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Registry bucket when no household profile is selected yet — downloads still run locally. */
internal const val OFFLINE_LOCAL_MEMBER_KEY = "__local__"

/** Server-synced registry of offline collections per household profile. */
@Serializable
data class OfflineDownloadRecord(
    val id: String,
    val title: String,
    val kind: String = "playlist",
    val sourcePlaylistId: String? = null,
)

object OfflineDownloadSync {
    private const val PREFS = "offline_download_registry"
    private const val KEY = "by_member"
    private val json = Json { ignoreUnknownKeys = true }

    fun register(context: Context, target: PlayTarget) {
        add(context, target.toOfflineDownloadRecord())
    }

    fun remove(context: Context, collectionId: String) {
        val memberId = registryMemberId(context)
        val map = loadMap(context).toMutableMap()
        val list = map[memberId]?.toMutableList() ?: return
        if (list.removeAll { it.id == collectionId }) {
            map[memberId] = list
            saveMap(context, map)
        }
    }

    /** Records for the active profile (server source of truth after pull). */
    fun recordsForActiveProfile(context: Context): List<OfflineDownloadRecord> {
        val memberId = ActiveProfileStore.activeMemberId(context) ?: return emptyList()
        return loadMap(context)[memberId].orEmpty()
    }

    /** Collection ids visible for the active profile (or local bucket before profile pick). */
    fun visibleCollectionIds(context: Context): Set<String> =
        collectionIdsForMember(context, registryMemberId(context))

    fun collectionIdsForMember(context: Context, memberId: String?): Set<String> {
        if (memberId.isNullOrBlank()) return emptySet()
        return loadMap(context)[memberId]?.map { it.id }?.toSet() ?: emptySet()
    }

    fun collectForMember(context: Context): List<OfflineDownloadRecord> =
        recordsForMember(context, ActiveProfileStore.activeMemberId(context))

    fun recordsForMember(context: Context, memberId: String?): List<OfflineDownloadRecord> {
        if (memberId.isNullOrBlank()) return emptyList()
        return loadMap(context)[memberId].orEmpty()
    }

    /** Attach on-disk collections not yet assigned to any profile (legacy / pre-registry). */
    fun claimOrphansForActiveProfile(context: Context) {
        val memberId = ActiveProfileStore.activeMemberId(context) ?: return
        val map = loadMap(context).toMutableMap()
        val assigned = map.values.flatten().map { it.id }.toSet()
        val orphans = OfflineDownloadStore(context).listManifests()
            .filter { it.id !in assigned }
            .map { it.toRecord() }
        val localPending = map.remove(OFFLINE_LOCAL_MEMBER_KEY).orEmpty()
            .filter { it.id !in assigned }
        val toClaim = mergeRecords(orphans + localPending)
        if (toClaim.isEmpty()) return
        val list = map[memberId]?.toMutableList() ?: mutableListOf()
        list.addAll(toClaim)
        map[memberId] = mergeRecords(list)
        saveMap(context, map)
    }

    fun applyRemote(context: Context, records: List<OfflineDownloadRecord>) {
        val memberId = ActiveProfileStore.activeMemberId(context)
        if (!memberId.isNullOrBlank()) {
            val map = loadMap(context).toMutableMap()
            map[memberId] = mergeRecords(records)
            saveMap(context, map)
        }
        restoreMissing(context, records)
    }

    fun restoreMissing(context: Context, records: List<OfflineDownloadRecord>) {
        if (records.isEmpty()) return
        val appContext = context.applicationContext
        val store = OfflineDownloadStore(appContext)
        OfflineDownloadManager.refresh(appContext)
        for (record in records) {
            val target = record.toPlayTarget()
            val manifest = store.readManifest(record.id)
            if (manifest != null && store.isCollectionComplete(manifest)) continue
            if (OfflineDownloadManager.statusFor(context, target)?.state == DownloadState.Downloading) continue
            // Partial or failed downloads require an explicit user retry.
            if (OfflineDownloadManager.statusFor(context, target)?.state == DownloadState.Failed) continue
            if (manifest != null && !store.isCollectionComplete(manifest)) continue
            OfflineDownloadManager.download(appContext, target)
        }
    }

    fun encode(records: List<OfflineDownloadRecord>): String =
        json.encodeToString(records)

    fun decode(raw: String?): List<OfflineDownloadRecord>? {
        if (raw.isNullOrBlank()) return null
        return runCatching { json.decodeFromString<List<OfflineDownloadRecord>>(raw) }.getOrNull()
    }

    /** Clears registry prefs and decode cache (instrumented tests only). */
    internal fun resetRegistryForTests(context: Context) {
        cachedRaw = null
        cachedMap = emptyMap()
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .clear()
            .commit()
    }

    private fun registryMemberId(context: Context): String =
        ActiveProfileStore.activeMemberId(context) ?: OFFLINE_LOCAL_MEMBER_KEY

    private fun add(context: Context, record: OfflineDownloadRecord) {
        val memberId = registryMemberId(context)
        val map = loadMap(context).toMutableMap()
        val list = map[memberId]?.toMutableList() ?: mutableListOf()
        list.removeAll { it.id == record.id }
        list.add(record)
        map[memberId] = list
        saveMap(context, map)
    }

    private fun mergeRecords(records: List<OfflineDownloadRecord>): List<OfflineDownloadRecord> {
        val out = linkedMapOf<String, OfflineDownloadRecord>()
        records.forEach { out[it.id] = it }
        return out.values.toList()
    }

    // Decode cache — loadMap is called from composition on every download progress
    // tick; re-parsing the JSON registry each time janks the main thread.
    @Volatile private var cachedRaw: String? = null
    @Volatile private var cachedMap: Map<String, List<OfflineDownloadRecord>> = emptyMap()

    private fun loadMap(context: Context): Map<String, List<OfflineDownloadRecord>> {
        val raw = context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY, null)
            ?: return emptyMap()
        if (raw == cachedRaw) return cachedMap
        val parsed = runCatching {
            json.decodeFromString<Map<String, List<OfflineDownloadRecord>>>(raw)
        }.getOrDefault(emptyMap())
        cachedRaw = raw
        cachedMap = parsed
        return parsed
    }

    private fun saveMap(context: Context, map: Map<String, List<OfflineDownloadRecord>>) {
        val encoded = json.encodeToString(map)
        cachedRaw = encoded
        cachedMap = map
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY, encoded)
            .apply()
    }
}

fun PlayTarget.toOfflineDownloadRecord(): OfflineDownloadRecord = OfflineDownloadRecord(
    id = downloadId(),
    title = label,
    kind = downloadKindLabel().lowercase(),
    sourcePlaylistId = (this as? PlayTarget.Playlist)?.id,
)

private fun OfflineDownloadRecord.toPlayTarget(): PlayTarget = when (kind) {
    "artist" -> PlayTarget.Artist(title)
    "album" -> PlayTarget.Album(title)
    "song" -> PlayTarget.Song(path = "", title = title)
    "mix", "radio" -> PlayTarget.Radio(title, PlayTarget.RadioSeedKind.Artist, title)
    else -> {
        val pid = sourcePlaylistId ?: id.removePrefix("pl-")
        PlayTarget.Playlist(pid, title)
    }
}

fun OfflineCollectionManifest.toRecord(): OfflineDownloadRecord = OfflineDownloadRecord(
    id = id,
    title = title,
    kind = kind,
    sourcePlaylistId = sourcePlaylistId,
)

File: ios/BockMedia/Offline/OfflineDownloadSync.swift
import Foundation

struct OfflineDownloadRecord: Codable, Equatable {
    var id: String
    var title: String
    var kind: String = "playlist"
    var sourcePlaylistId: String?
}

enum OfflineDownloadSync {
    private static let defaultsKey = "offlineDownloadRegistryByMember"

    static func register(target: PlayTarget) {
        add(target.toOfflineDownloadRecord())
    }

    static func remove(collectionId: String) {
        guard let memberId = ActiveProfileStore.activeMemberId() else { return }
        var map = loadMap()
        var list = map[memberId] ?? []
        list.removeAll { $0.id == collectionId }
        map[memberId] = list
        saveMap(map)
    }

    static func visibleCollectionIds() -> Set<String> {
        guard let memberId = ActiveProfileStore.activeMemberId() else { return [] }
        return Set((loadMap()[memberId] ?? []).map(\.id))
    }

    /// Attach on-disk collections not yet assigned to any profile (legacy / pre-registry).
    static func claimOrphansForActiveProfile(store: OfflineDownloadStore = OfflineDownloadStore()) {
        guard let memberId = ActiveProfileStore.activeMemberId() else { return }
        var map = loadMap()
        let assigned = Set(map.values.flatMap { $0 }.map(\.id))
        let orphans = store.listManifests()
            .filter { !assigned.contains($0.id) }
            .map { $0.toRecord() }
        guard !orphans.isEmpty else { return }
        var list = map[memberId] ?? []
        list.append(contentsOf: orphans)
        map[memberId] = mergeRecords(list)
        saveMap(map)
    }

    static func collectionIdsForMember(_ memberId: String) -> [String] {
        (loadMap()[memberId] ?? []).map(\.id)
    }

    static func collectForMember(store: OfflineDownloadStore = OfflineDownloadStore()) -> [OfflineDownloadRecord] {
        guard let memberId = ActiveProfileStore.activeMemberId() else { return [] }
        let fromRegistry = loadMap()[memberId] ?? []
        let fromDisk = store.listManifests().map { $0.toRecord() }
        return mergeRecords(fromRegistry + fromDisk)
    }

    static func applyRemote(_ records: [OfflineDownloadRecord], repository: BockMediaRepository) {
        if let memberId = ActiveProfileStore.activeMemberId() {
            var map = loadMap()
            map[memberId] = mergeRecords(records)
            saveMap(map)
        }
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 400_000_000)
            restoreMissing(records, repository: repository)
        }
    }

    @MainActor
    static func restoreMissing(_ records: [OfflineDownloadRecord], repository: BockMediaRepository) {
        guard !records.isEmpty else { return }
        let store = OfflineDownloadStore()
        let manager = OfflineDownloadManager.shared
        manager.refresh()
        let prefs = repository.preferences
        for record in records {
            let target = record.toPlayTarget()
            if let manifest = store.readManifest(record.id), store.isCollectionComplete(manifest) { continue }
            if manager.status(for: target)?.state == .downloading { continue }
            manager.download(repository: repository, preferences: prefs, target: target)
        }
    }

    static func encode(_ records: [OfflineDownloadRecord]) -> Any? {
        guard let data = try? JSONEncoder().encode(records) else { return nil }
        return try? JSONSerialization.jsonObject(with: data)
    }

    static func decode(_ value: Any?) -> [OfflineDownloadRecord]? {
        guard let value else { return nil }
        guard let data = try? JSONSerialization.data(withJSONObject: value) else { return nil }
        return try? JSONDecoder().decode([OfflineDownloadRecord].self, from: data)
    }

    private static func add(_ record: OfflineDownloadRecord) {
        guard let memberId = ActiveProfileStore.activeMemberId() else { return }
        var map = loadMap()
        var list = map[memberId] ?? []
        list.removeAll { $0.id == record.id }
        list.append(record)
        map[memberId] = list
        saveMap(map)
    }

    private static func mergeRecords(_ records: [OfflineDownloadRecord]) -> [OfflineDownloadRecord] {
        var out: [String: OfflineDownloadRecord] = [:]
        records.forEach { out[$0.id] = $0 }
        return Array(out.values)
    }

    private static func loadMap() -> [String: [OfflineDownloadRecord]] {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey) else { return [:] }
        return (try? JSONDecoder().decode([String: [OfflineDownloadRecord]].self, from: data)) ?? [:]
    }

    private static func saveMap(_ map: [String: [OfflineDownloadRecord]]) {
        guard let data = try? JSONEncoder().encode(map) else { return }
        UserDefaults.standard.set(data, forKey: defaultsKey)
    }
}

private extension PlayTarget {
    func toOfflineDownloadRecord() -> OfflineDownloadRecord {
        OfflineDownloadRecord(
            id: downloadId(),
            title: label,
            kind: downloadKindLabel().lowercased(),
            sourcePlaylistId: {
                if case .playlist(let pid, _) = self { return pid }
                return nil
            }()
        )
    }
}

private extension OfflineDownloadRecord {
    func toPlayTarget() -> PlayTarget {
        switch kind {
        case "artist": return .artist(name: title)
        case "album": return .album(name: title, artist: nil)
        case "song": return .song(path: "", title: title)
        case "mix", "radio": return .radio(displayTitle: title, seedKind: .artist, name: title, path: nil)
        default:
            let pid = sourcePlaylistId ?? id.replacingOccurrences(of: "pl-", with: "")
            return .playlist(id: pid, name: title)
        }
    }
}

private extension OfflineCollectionManifest {
    func toRecord() -> OfflineDownloadRecord {
        OfflineDownloadRecord(id: id, title: title, kind: kind, sourcePlaylistId: sourcePlaylistId)
    }
}
